osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObjectContainer.cs

```diff
@@ -0,0 +1,232 @@
+        private readonly DrawablePool<DrawableCaughtDroplet> caughtDropletPool;
+
+        /// <summary>
+        /// The randomness used to compute position in stack.
+        /// It is incremented for each <see cref="Add"/> call and decremented for each <see cref="Remove"/> call to make a replay consistent.
+        /// </summary>
+        private int randomSeed = 1;
+
+        public CaughtObjectContainer(Container<DrawableCaughtObject> droppedObjectTarget)
+        {
+            this.droppedObjectTarget = droppedObjectTarget;
+
+            InternalChildren = new Drawable[]
+            {
+                caughtFruitPool = new DrawablePool<DrawableCaughtFruit>(50),
+                caughtBananaPool = new DrawablePool<DrawableCaughtBanana>(100),
+                // less capacity is needed compared to fruit because droplet is not stacked
+                caughtDropletPool = new DrawablePool<DrawableCaughtDroplet>(25),
+                StackedObjectContainer = new Container<DrawableCaughtObject>()
+            };
+
+            lifetimeManager.EntryBecameAlive += entryBecameAlive;
+            lifetimeManager.EntryBecameDead += entryBecameDead;
+        }
+
+        public void Add(CaughtObjectEntry entry)
+        {
+            randomSeed++;
+
+            lifetimeManager.AddEntry(entry);
+
+            if (entry.State != CaughtObjectState.Stacked) return;
+
+            // `DropStackedObjects` may be called before lifetime update.
+            if (entry.LifetimeStart <= Time.Current)
+                aliveStackedObjects.Add(entry);
+
+            var dropEntry = new CaughtObjectEntry(CaughtObjectState.Dropped, entry.PositionInStack, entry)
+            {
+                LifetimeStart = double.PositiveInfinity,
+                LifetimeEnd = double.PositiveInfinity
+            };
+
+            lifetimeManager.AddEntry(dropEntry);
+            dropEntryMap[entry] = dropEntry;
+        }
+
+        public void Remove(CaughtObjectEntry entry)
+        {
+            randomSeed--;
+
+            lifetimeManager.RemoveEntry(entry);
+            removeDrawable(entry);
+
+            aliveStackedObjects.Remove(entry);
+
+            if (!dropEntryMap.TryGetValue(entry, out var dropEntry)) return;
+
+            dropEntryMap.Remove(entry);
+
+            lifetimeManager.RemoveEntry(dropEntry);
+            removeDrawable(dropEntry);
+        }
+
+        public Vector2 GetPositionInStack(Vector2 position, float displayRadius)
+        {
+            const float radius_div_2 = CatchHitObject.OBJECT_RADIUS / 2;
+            const float allowance = 10;
+
+            int iteration = 0;
+
+            while (aliveStackedObjects.Any(f => Vector2Extensions.Distance(f.PositionInStack, position) < (displayRadius + radius_div_2) / (allowance / 2)))
+            {
+                float diff = (displayRadius + radius_div_2) / allowance;
+
+                position.X += (StatelessRNG.NextSingle(randomSeed, iteration + 2) - 0.5f) * diff * 2;
```

Comment:
guessing sequential values were intended here

```suggestion
                position.X += (StatelessRNG.NextSingle(randomSeed, iteration * 2) - 0.5f) * diff * 2;
```